feat(config): scoped secrets using pgp/gpg (#11673)
renovatebot/renovate: docs/usage/self-hosted-configuration.md

Any encrypted secrets using GPG must have a mandatory organization/group scope, and optionally can be scoped for a single repository only.
The reason for this is to avoid "replay" attacks where someone could learn your encrypted secret and then reuse it in their own Renovate repositories.
Instead, with scoped secrets it means that Renovate ensures that the organization and optionally repository values encrypted with the secret match against the running repository.

Note: simple public key encryption was previously used to encrypt secrets, but this approach has now been deprecated and no longer documented.

## privateKeyOld

Use this field if you need to perform a "key rotation" and support more than one keypair at a time.
Decryption with this key will be attempted after `privateKey`.

If you are migrating from the legacy public key encryption approach to use GPG, then move your legacy private key from `privateKey` to `privateKeyOld` and then put your new GPG private key in `privateKey`.
Doing so will mean that Renovate will first attempt to decrypt using the GPG key but fall back to the legacy key and try that next.

## privateKeyPath

> Used as an alternative to `privateKey`, if you wish for the key to be read from disk instead.

Reviewer: ```suggestion
Used as an alternative to `privateKey`, if you want the key to be read from disk instead.
```

## privateKeyPathOld

> Used as an alternative to `privateKeyOld`, if you wish for the key to be read from disk instead.

Reviewer: ```suggestion
Used as an alternative to `privateKeyOld`, if you want the key to be read from disk instead.
```

## productLinks

> Override this object if you wish to change the URLs that Renovate links to, e.g. if you have an internal forum for asking for help.

Reviewer: ```suggestion
Override this object if you want to change the URLs that Renovate links to, e.g. if you have an internal forum for asking for help.
```

> The corresponding public key can be used to create encrypted values for config files.

Reviewer: The rich diff preview is failing, I'm not sure what the reason for that is...

Author: Works for me

Reviewer: It's working for me now as well. 😄 Weird...

Your selection? 1
RSA keys may be between 1024 and 4096 bits long.
What keysize do you want? (3072) 4096
Requested keysize is 4096 bits
Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years
Key is valid for? (0)
Key does not expire at all
Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.

Real name: Renovate Bot
Email address: renovate@whitesourcesoftware.com
Comment:
You selected this USER-ID:
    "Renovate Bot <renovate@whitesourcesoftware.com>"

Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O
We need to generate a lot of random bytes. It is a good idea to perform
some other action (type on the keyboard, move the mouse, utilize the
disks) during the prime generation; this gives the random number
generator a better chance to gain enough entropy.
We need to generate a lot of random bytes. It is a good idea to perform
some other action (type on the keyboard, move the mouse, utilize the
disks) during the prime generation; this gives the random number

> generator a better chance to gain enough entropy.

Reviewer: The "do things on your keyboard/mouse" part is shown twice, is that intentional?

Author: It's from the log but I think we can remove it as it's not important

gpg: key 0649CC3899F22A66 marked as ultimately trusted
gpg: revocation certificate stored as '/Users/rhys/.gnupg/openpgp-revocs.d/794B820F34B34A8DF32AADB20649CC3899F22A66.rev'
public and secret key created and signed.

pub   rsa4096 2021-09-10 [SC]
      794B820F34B34A8DF32AADB20649CEXAMPLEONLY
uid                      Renovate Bot <renovate@whitesourcesoftware.com>
sub   rsa4096 2021-09-10 [E]
```

</details>

- Copy the key ID from the output (`794B820F34B34A8DF32AADB20649CEXAMPLEONLY` in the above example) or run `gpg --list-secret-keys` if you forgot to take a copy
- Run `gpg --armor --export-secret-keys YOUR_NEW_KEY_ID > renovate-private-key.asc` to generate an armored (text-based) private key file
- Run `gpg --armor --export YOUR_NEW_KEY_ID > renovate-public-key.asc` to generate an armored (text-based) public key file

The private key should then be added to your Renovate Bot global config (either using `privateKeyPath` or exporting it to the `RENOVATE_PRIVATE_KEY` environment variable).
The public key can be used to replace the existing key in <https://renovatebot.com/encrypt> for your own use.

> Any encrypted secrets using PGP must have a mandatory organization/group scope, and optionally can be scoped for a single repository only.

Reviewer: ```suggestion
Any encrypted secrets using GPG must have a mandatory organization/group scope, and optionally can be scoped for a single repository only.
```

Not sure if you mean PGP or GPG... Confusing naming!

The reason for this is to avoid "replay" attacks where someone could learn your encrypted secret and then reuse it in their own Renovate repositories.
Instead, with scoped secrets it means that Renovate ensures that the organization and optionally repository values encrypted with the secret match against the running repository.

Note: simple public key encryption was previously used to encrypt secrets, but this approach has now been deprecated and no longer documented.

## privateKeyOld

Use this field if you need to perform a "key rotation" and support more than one keypair at a time.
Decryption with this key will be attempted after `privateKey`.

> If you are migrating from the legacy public key encryption approach to use PGP, then move your legacy private key from `privateKey` to `privateKeyOld` and then put your new PGP private key in `privateKey`.
> Doing so will mean that Renovate will first attempt to decrypt using the PGP key but fall back to the legacy key and try that next.

Reviewer: ```suggestion
If you are migrating from the legacy public key encryption approach to use GPG, then move your legacy private key from `privateKey` to `privateKeyOld` and then put your new GPG private key in `privateKey`.
Doing so will mean that Renovate will first attempt to decrypt using the GPG key but fall back to the legacy key and try that next.
```

Again, do we mean PGP or GPG here? Confusing... 😄

The corresponding public key can be used to create encrypted values for config files.

> If you want a simple UI to encrypt values you can put the public key in a HTML page similar to <https://renovatebot.com/encrypt>.

Reviewer: PR title says: `pgp`
In docs we use `gpg`

Not sure which PR title is actually correct here...

To create the key pair with GPG use the following commands:

- `gpg --full-generate-key` and follow the prompts to generate a key. Name and email are not important to Renovate, and do not configure a passphrase. Use a 4096bit key.

<details><summary>key generation log</summary>

```
❯ gpg --full-generate-key
gpg (GnuPG) 2.2.24; Copyright (C) 2020 Free Software Foundation, Inc.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Please select what kind of key you want:
   (1) RSA and RSA (default)
   (2) DSA and Elgamal
   (3) DSA (sign only)
   (4) RSA (sign only)
  (14) Existing key from card
Your selection? 1
RSA keys may be between 1024 and 4096 bits long.
What keysize do you want? (3072) 4096
Requested keysize is 4096 bits
Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years

> Key is valid for? (0)

Reviewer: Is it a smart idea to make a Renovate bot key that doesn't expire? I thought usually its a good idea to rotate keys like this? But maybe it's really inconvenient if the bot suddenly stops working because the key expired on a set schedule?

You probably know better than me what is a good idea and what isn't, so I'll let you decide. 😉

Author: It's ok if they want the key to expire. Ref: https://community.progress.com/s/article/Can-a-PGP-encryption-key-be-used-after-the-expiration-date

The public key can be used to replace the existing key in <https://renovatebot.com/encrypt> for your own use.

Any encrypted secrets using GPG must have a mandatory organization/group scope, and optionally can be scoped for a single repository only.
The reason for this is to avoid "replay" attacks where someone could learn your encrypted secret and then reuse it in their own Renovate repositories.
Instead, with scoped secrets it means that Renovate ensures that the organization and optionally repository values encrypted with the secret match against the running repository.

Note: simple public key encryption was previously used to encrypt secrets, but this approach has now been deprecated and no longer documented.

## privateKeyOld

Use this field if you need to perform a "key rotation" and support more than one keypair at a time.
Decryption with this key will be attempted after `privateKey`.

If you are migrating from the legacy public key encryption approach to use GPG, then move your legacy private key from `privateKey` to `privateKeyOld` and then put your new GPG private key in `privateKey`.

> Doing so will mean that Renovate will first attempt to decrypt using the GPG key but fall back to the legacy key and try that next.

Reviewer: I'm missing an explanation on when it's safe to remove the `privateKeyOld` config option when doing the "key rotation".

Maybe we want a user to do something like this?

1. Create new GPG key.
1. Put old key in `privateKeyOld` config option.
1. Confirm new GPG key works (how would you even confirm this if the old key keeps hanging around, though...)
1. Maybe wait until the PRs which use `privateOldKey` have been merged???
1. Remove `privateKeyOld` config option.

I'll let you respond and decide if this needs to go in the text. 😉

Author: They can remove it once all the old encrypted values have been migrated, or they wish to no longer support the old key and let repos fail

Requested keysize is 4096 bits
Please specify how long the key should be valid.
         0 = key does not expire
      <n>  = key expires in n days
      <n>w = key expires in n weeks
      <n>m = key expires in n months
      <n>y = key expires in n years
Key is valid for? (0)
Key does not expire at all
Is this correct? (y/N) y

GnuPG needs to construct a user ID to identify your key.

Real name: Renovate Bot
Email address: renovate@whitesourcesoftware.com
Comment:
You selected this USER-ID:
    "Renovate Bot <renovate@whitesourcesoftware.com>"

Change (N)ame, (C)omment, (E)mail or (O)kay/(Q)uit? O

gpg: key 0649CC3899F22A66 marked as ultimately trusted
gpg: revocation certificate stored as '/Users/rhys/.gnupg/openpgp-revocs.d/794B820F34B34A8DF32AADB20649CC3899F22A66.rev'
public and secret key created and signed.

pub   rsa4096 2021-09-10 [SC]
      794B820F34B34A8DF32AADB20649CEXAMPLEONLY
uid                      Renovate Bot <renovate@whitesourcesoftware.com>
sub   rsa4096 2021-09-10 [E]
```

</details>

- Copy the key ID from the output (`794B820F34B34A8DF32AADB20649CEXAMPLEONLY` in the above example) or run `gpg --list-secret-keys` if you forgot to take a copy

> - Run `gpg --armor --export-secret-keys YOUR_NEW_KEY_ID > renovate-private-key.asc` to generate an armored (text-based) private key file

Author: Yes, good point

> - Run `gpg --armor --export-secret-keys YOUR_NEW_KEY_ID > renovate-private-key.asc` to generate an armored (text-based) private key file

Author: Actually, the description above already says "do not configure a passphrase"
